### ISSUE TITLE
mirobo updater -- No request was made

### ISSUE THREAD

**Issue by chrisx51** [rytilahti/python-miio#557]
Hello all,
I am investing my 3rd evening on trying to get valetudo running. I got completely stuck using virtualbox and vmware, but now by using my raspberry I think I am nearly there.
I managed to read the token alreay.
But when trying to update the following happens:

```
(venv) pi@raspberrypi2:~/flash $ mirobo --ip 192.168.8.1 --token xxxxxxxxx update-firmware v11_003468.pkg
Going to update from v11_003468.pkg
INFO:miio.updater:Serving on 0.0.0.0:35709, timeout 10
INFO:miio.updater:Using local v11_003468.pkg (md5: 65688e671050ac168a79975f2fa15aaa)
Hosting file at http://192.168.178.38:35709/v11_003468.pkg
Update started!
  0%|                                                   | 0/100 [00:05<?, ?it/s]
**ERROR:miio.updater:No request was made..**
```
Every help is highly appreciated...

**Comment by rytilahti** [comment]
You need to use `--ip` of `update-firmware` to force another IP address for updates (I assume that there is no routing between networks 192.168.8.0/24 and 192.168.178.0/24), so the command should be `mirobo --ip 192.168.8.1 --token xxxxxxxxx update-firmware --ip 192.168.8.<YOUR ADDR IN VACUUM NET> v11_003468.pkg`. 

You can use `--help` with any command to get command-specific help, e.g., `mirobo update-firmware --help` show the help for update-firmware.

**Comment by chrisx51** [comment]
@rytilahti Thank you so much for your hint! This now saves my weekend, as now it is working!
It is always great to see that there are people outside in the world to support.

**Comment by zackdvd** [comment]
Hey, thanks for your help, but this isn't working for me. 

`$ mirobo --ip 192.168.8.1 --token 7342514b7941584e4d736d796xxxxxxx update-firmware v11_003468.pkg
Going to update from v11_003468.pkg
INFO:miio.updater:Serving on 0.0.0.0:50369, timeout 10
INFO:miio.updater:Using local v11_003468.pkg (md5: 58f985c855db37104bb955118ad268be)
Hosting file at http://192.168.8.2:50369/v11_003468.pkg
Update started!
  0%|                                                                                                               | 0/100 [00:05<?, ?it/s]
ERROR:miio.updater:No request was made..`

Any ideas on that? Host is macos 10.14, if that is important. 
I also did WIFI reset, not help. Then I did complete reset, no help. But the vacuum now speaks something in chinese before error message appears...

**Comment by thrasheri** [comment]
Hi!

I'm facing the exact same issue, as @zackdvd  but on Windows 10 host. I also did wifi reset and factory reset. The update starts, but doesn't proceed from 0%

**Comment by rytilahti** [comment]
Did you try to specify the correct IP address to the command like shown in my last comment (assuming you have multiple network interfaces)?

**Comment by thrasheri** [comment]
I had the ip-address specified. I got my issue resolved by disabling all the other network interfaces.

**Comment by zackdvd** [comment]
Hi, finally it also worked for me. Somehow a reboot of my mac did the thing (maybe some processes before did not stop running?), after that some different error messages but 3rd attempt was fine!

Please try to add --debug and see if you can find any error.

Good luck!

**Comment by dd23** [comment]
I also got `ERROR:miio.updater:No request was made` both with and without `--ip 192.168.8.<YOUR ADDR IN VACUUM NET>` after update-firmware.

What eventually helped was:
 - not specifying my device IP (it was detected automatically)
 - running the command with the debug flag

So the command I ran, that worked was:
`mirobo --ip 192.168.8.1 --token 12345xxxxxxxxx -d update-firmware v11_003468.pkg`

**Comment by luxus** [comment]
hello, i have the same issue but no fix so far :D
```
mirobo --ip 192.168.8.1 --token '594234586238316f39476b6c43526f70'                                                                                                                                                                        
State: Idle
Battery: 98 %
Fanspeed: 60 %
Cleaning since: 0:00:00
```
worked
````
mirobo --ip 192.168.8.1 --token 594234586238316f39476b6c43526f70 -d update-firmware firmware/output/v11_003468.fullos.pkg                                                                                                                 
INFO:miio.vacuum_cli:Debug mode active
DEBUG:miio.vacuum_cli:Read stored sequence ids: {'seq': 28, 'manual_seq': 0}
DEBUG:miio.vacuum_cli:Connecting to 192.168.8.1 with token 594234586238316f39476b6c43526f70
Going to update from firmware/output/v11_003468.fullos.pkg
INFO:miio.updater:Serving on 0.0.0.0:43869, timeout 10
INFO:miio.updater:Using local firmware/output/v11_003468.fullos.pkg (md5: ec9df4bfa90f94f784bd505fee9b261d)
DEBUG:miio.updater:available interfaces: ['enp6s0', 'enp0s31f6', 'wlp0s20f3']
DEBUG:miio.updater:enp6s0 has no ipv4 addresses, skipping
DEBUG:miio.updater:enp0s31f6 has no ipv4 addresses, skipping
DEBUG:miio.updater:Got addr: 192.168.8.67
Hosting file at http://192.168.8.67:43869/v11_003468.fullos.pkg
DEBUG:miio.protocol:Unable to decrypt, returning raw bytes: b''
DEBUG:miio.miioprotocol:Got a response: Container: 
    data = Container: 
        data = b'' (total 0)
        value = b'' (total 0)
        offset1 = 32
        offset2 = 32
        length = 0
    header = Container: 
        data = b'!1\x00 \x00\x00\x00\x00\x02\xf2n\xc9\x00\x00\x03\xe9' (total 16)
        value = Container: 
            length = 32
            unknown = 0
            device_id = unhexlify('02f26ec9')
            ts = 1970-01-01 00:16:41
        offset1 = 0
        offset2 = 16
        length = 16
    checksum = b'YB4Xb81o9GklCRop' (total 16)
DEBUG:miio.miioprotocol:Discovered 02f26ec9 with ts: 1970-01-01 00:16:41, token: b'594234586238316f39476b6c43526f70'
DEBUG:miio.miioprotocol:192.168.8.1:54321 >>: {'id': 29, 'method': 'miIO.ota', 'params': {'mode': 'normal', 'install': '1', 'app_url': 'http://192.168.8.67:43869/v11_003468.fullos.pkg', 'file_md5': 'ec9df4bfa90f94f784bd505fee9b261d', 'proc': 'dnld install'}}
DEBUG:miio.miioprotocol:192.168.8.1:54321 (ts: 1970-01-01 00:16:41, id: 29) << {'result': ['ok'], 'id': 29}
Update started!
  0%|                                                                                                                                                                                                                               | 0/100 [00:00<?, ?it/s]DEBUG:miio.miioprotocol:192.168.8.1:54321 >>: {'id': 30, 'method': 'miIO.get_ota_state', 'params': []}
DEBUG:miio.miioprotocol:192.168.8.1:54321 (ts: 1970-01-01 00:16:41, id: 30) << {'result': ['failed'], 'id': 30}
  0%|                                                                                                                                                                                                                               | 0/100 [00:00<?, ?it/s]
DEBUG:miio.vacuum_cli:Writing {'seq': 30, 'manual_seq': 0} to /home/luxus/.cache/python-miio/python-mirobo.seq
ERROR:miio.updater:No request was made..
```
failed.. i have a very very old v1, i did a hard reset .. i'm afraid that the software is too old

**Comment by rytilahti** [comment]
I'm not sure what could cause it not to work with statically defined address while still working with the automatically detected one (as that does not make much sense), but I'm glad it worked for you @dd23!

@luxus I suppose 192.168.8.67 the IP address of your host system? Is there potentially a firewall blocking the incoming requests? If you are comfortable with it, you could try to create a traffic capture with tcpdump (or wireshark) to see if the host receives a request for that file.

**Comment by luxus** [comment]
I got it working... it was pretty easy and im sorry for wasting ur time...
I didn't charge it while trying to update. Now i sending the firmware right in this moment

**Comment by dvbit** [comment]
I have been struggling with the same error v1 robo:
- have the robo docked
. reset firmware to factory: Home + reset pin until lights off. Then keep holding Home. Wait for China Girl to say it's resetting.Go with the command above.  Wait a few minutes (be patient). Success

**Comment by Spartan-II-117** [comment]
hello, I have tried repeated factory resets and am still not getting anywhere. here is my debug output of the session:

DEBUG:miio.miioprotocol:192.168.8.1:54321 >>: {'id': 1350, 'method': 'miIO.ota', 'params': {'mode': 'normal', 'install': '1', 'app_url': 'http://192.168.8.62:41981/v11_004018.pkg', 'file_md5': '14c89c1e01112b493414f3b046b214e9', 'proc': 'dnld install'}}
DEBUG:miio.miioprotocol:192.168.8.1:54321 (ts: 1970-01-01 00:01:50, id: 1350) << {'result': ['ok'], 'id': 1350}
Update started!
  0%|                                                                                                                                          | 0/100 [00:00<?, ?it/s]DEBUG:miio.miioprotocol:192.168.8.1:54321 >>: {'id': 1351, 'method': 'miIO.get_ota_state', 'params': []}
DEBUG:miio.miioprotocol:Retrying with incremented id, retries left: 3
DEBUG:miio.protocol:Unable to decrypt, returning raw bytes: b''
DEBUG:miio.miioprotocol:Got a response: Container: 
    data = Container: 
        data = b'' (total 0)
        value = b'' (total 0)
        offset1 = 32
        offset2 = 32
        length = 0
    header = Container: 
        data = b'!1\x00 \x00\x00\x00\x00\x04l\x9c\xe5\x00\x00\x00s' (total 16)
        value = Container: 
            length = 32
            unknown = 0
            device_id = unhexlify('046c9ce5')
            ts = 1970-01-01 00:01:55
        offset1 = 0
        offset2 = 16
        length = 16
    checksum = b'NDIYyaNQmcYEY3H6' (total 16)
DEBUG:miio.miioprotocol:Discovered 046c9ce5 with ts: 1970-01-01 00:01:55, token: b'4e44495979614e516d63594559334836'
DEBUG:miio.miioprotocol:192.168.8.1:54321 >>: {'id': 1452, 'method': 'miIO.get_ota_state', 'params': []}
DEBUG:miio.miioprotocol:192.168.8.1:54321 (ts: 1970-01-01 00:01:55, id: 1452) << {'result': ['failed'], 'id': 1452}
  0%|                                                                                                                                          | 0/100 [00:05<?, ?it/s]
DEBUG:miio.vacuum_cli:Writing {'seq': 1452, 'manual_seq': 0} to /home/user/.cache/python-miio/python-mirobo.seq
ERROR:miio.updater:No request was made..

**Comment by nicbou** [comment]
**Problem**

The update would start, then the vacuum would say something in Chinese, and I would get `ERROR:miio.updater:No request was made..`

**Solution**

1. **Put the vacuum in the dock!**
2. Make sure it's on (the power button light should be on)
3. Run this command: `mirobo --ip 192.168.8.1 --token YOUR_TOKEN_HERE --debug update-firmware YOUR_PACKAGE_FILE_NAME.pkg`

That's it! You should see the progress go up (and lots of debug messages). When it's done, the vacuum will say something in Chinese, and the power button will blink quickly. Wait a bit, and you're good to go.

**Comment by rytilahti** [comment]
There are likely some prereqs from the device (docked, enough battery left, ..?) before the device starts an update. Glad you got it working!

Anyway, feel free to extend https://github.com/rytilahti/python-miio/blob/master/docs/device_docs/vacuum.rst#firmware-update - it's more visible place than the isse comments :-)

**Comment by horror-vacui** [comment]
I stuck with the same problem for quiet a while... The turning point come, when I started to suspect, that even though I tried pressing the reset button and nothing happened, I should factory reset the vacuum. After some googling I found out the not so trivial factory reset sequence: hold the home button, press the reset button, keep it pressed until the lights go away, release the reset button, but keep holding the house button, until the robot does not announce that it resets itself.
After that I was able to flash the firmware.

I hope it will save some headache for those who come later...
--------------------------------------------------------------------
The process I've followed:

0. charge the vacuum and keep it in the dock
1. generate ssh keypair (for later ssh access of the vacuum; not related for the firmware flashing)
2. generate a firmware image with dustcloud
3. reset the wifi in the vacuum (holding the power and house button down for approx. 5 seconds)
4. connect to the wifi of the vacuum. Ethernet cable is unconnected (no other network).
5. get the token information with `mirobo --debug discover --handshake true`; The token will change after every wifi reset!
6. update the firmware with `mirobo --ip 192.168.8.1 --token 123465.....123456789 --debug update-firmware disk.img`

If it still does not work, make a factory reset.
